internal/controller/promotion/argocd.go

```diff
@@ -233,76 +275,39 @@ func (a *argoCDMechanism) mustPerformUpdate(
 		)
 	}
 
+	// Check if the desired source(s) were applied.
+	if len(update.SourceUpdates) > 0 {
+		if (desiredSource != nil && !desiredSource.Equals(&status.SyncResult.Source)) ||
+			!desiredSources.Equals(status.SyncResult.Sources) {
+			// The operation did not result in the desired source(s) being applied.
+			// We should attempt to retry the operation.
+			return "", true, fmt.Errorf(
+				"operation result source does not match desired source",
+			)
+		}
+	}
+
 	// The operation has completed.
 	return status.Phase, false, nil
 }
 
-func (a *argoCDMechanism) doSingleUpdate(
+func (a *argoCDMechanism) updateApplicationSources(
 	ctx context.Context,
-	stageMeta metav1.ObjectMeta,
-	update kargoapi.ArgoCDAppUpdate,
-	newFreight kargoapi.FreightReference,
+	app *argocd.Application,
+	desiredSource *argocd.ApplicationSource,
+	desiredSources argocd.ApplicationSources,
 ) error {
-	namespace := update.AppNamespace
-	if namespace == "" {
-		namespace = libargocd.Namespace()
-	}
-	app, err := a.getArgoCDAppFn(ctx, namespace, update.AppName)
-	if err != nil {
-		return fmt.Errorf(
-			"error finding Argo CD Application %q in namespace %q: %w",
-			update.AppName,
-			namespace,
-			err,
-		)
-	}
-	if app == nil {
-		return fmt.Errorf(
-			"unable to find Argo CD Application %q in namespace %q: %w",
-			update.AppName,
-			namespace,
-			err,
-		)
-	}
-	// Make sure this is allowed!
-	if err = authorizeArgoCDAppUpdate(stageMeta, app.ObjectMeta); err != nil {
-		return err
-	}
+	// Create a patch for the Application.
 	patch := client.MergeFrom(app.DeepCopy())
-	for _, srcUpdate := range update.SourceUpdates {
-		if app.Spec.Source != nil {
-			var source argocd.ApplicationSource
-			if source, err = a.applyArgoCDSourceUpdateFn(
-				*app.Spec.Source,
-				newFreight,
-				srcUpdate,
-			); err != nil {
-				return fmt.Errorf(
-					"error updating source of Argo CD Application %q in namespace %q: %w",
-					update.AppName,
-					namespace,
-					err,
-				)
-			}
-			app.Spec.Source = &source
-		}
-		for i, source := range app.Spec.Sources {
-			if source, err = a.applyArgoCDSourceUpdateFn(
-				source,
-				newFreight,
-				srcUpdate,
-			); err != nil {
-				return fmt.Errorf(
-					"error updating source(s) of Argo CD Application %q in namespace %q: %w",
-					update.AppName,
-					namespace,
-					err,
-				)
-			}
-			app.Spec.Sources[i] = source
-		}
-	}
+
+	// Initiate a "hard" refresh.
 	app.ObjectMeta.Annotations[argocd.AnnotationKeyRefresh] = string(argocd.RefreshTypeHard)
+
+	// Update the desired source(s) in the Argo CD Application.
+	app.Spec.Source = desiredSource.DeepCopy()
+	app.Spec.Sources = desiredSources.DeepCopy()
```

Comment:
Just another spot where I am curious what the deep copies accomplished.

Reply:
This is a habit to ensure the Application fields do not unexpectedly become attached to the desired objects. Which may cause surprising bugs much later on when we suddenly do start to modify those things at some point.